File: remove-invalid-parentheses.js
// Time Complexity : O(2^N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

/**
 * @param {string} s
 * @return {string[]}
 */
var removeInvalidParentheses = function(s) {
    if (!s || !s.length) return [s];
    
    const isValid = (string) => {
        let count = 0;
        for (const char of string) {
            if (char == "(") count++;
            else if (char == ")") count--;
            if (count < 0) return false;
        }
        return count == 0;
    }

    if (isValid(s)) return [s];
    const result = new Set();
    const visited = new Set();
    const q = [s];
    let isDone = false;
    while (q.length > 0) {
        const levelCount = q.length;
        for (let i = 0; i < levelCount; i++) {
            let current = q.shift();
            if (!visited.has(current)) {
                if (isValid(current)) {
                    result.add(current);
                    isDone = true;
                }

                visited.add(current);

                if (!isDone) {
                    for (let j = 0; j < current.length; j++) {
                        if (current[j] == "(" || current[j] == ")") {
                            let temp = current.substring(0, j) + current.substring(j + 1);
                            q.push(temp);
                        }
                    }
                }
            }
        }
        if (isDone) break;
    }
    return Array.from(result);
};

File: RemoveInvalidParentheses.cs
using System;
using System.Collections.Generic;

namespace Algorithms
{
    public class RemoveInvalidParentheses
    {
        /// Time Complexity : O(N^N) 
        // Space Complexity :O(N^N) 
        // Did this code successfully run on Leetcode :Yes
        // Any problem you faced while coding this :  No 
        public IList<string> RemoveInvalidParentheses_BFS(string s)
        {
            IList<string> result = new List<string>();
            if (s == null || s.Length == 0) return result;
            Queue<string> q = new Queue<string>();
            HashSet<string> set = new HashSet<string>();
            q.Enqueue(s);
            bool flag = false;
            while (q.Count != 0)
            {

                string curr = q.Dequeue();
                if (IsValid(curr))
                {
                    result.Add(curr);
                    flag = true;
                }
                if (!flag)
                {
                    for (int j = 0; j < curr.Length; j++)
                    {
                        char[] charArray = curr.ToCharArray();
                        if (Char.IsLetter(charArray[j])) continue;
                        //int length = endIndex - startIndex + 1;                    
                        string sub1 = curr.Substring(0, j);
                        string sub2 = curr.Substring(j + 1, curr.Length - 1 - (j + 1) + 1);
                        string child = sub1 + sub2;
                        if (!set.Contains(child))
                        {
                            set.Add(child);
                            q.Enqueue(child);
                        }
                    }
                }
            }
            return result;
        }

        private bool IsValid(string s)
        {
            int counter = 0;
            char[] charArray = s.ToCharArray();
            for (int i = 0; i < s.Length; i++)
            {
                if (Char.IsLetter(charArray[i])) continue;
                if (charArray[i] == '(') counter++;
                else counter--;
                if (counter < 0) return false;
            }
            return counter == 0;
        }



        HashSet<string> set;
        IList<string> result;
        int max;
        public IList<string> RemoveInvalidParentheses_DFS(string s)
        {
            result = new List<string>();
            if (s == null || s.Length == 0) return result;
            set = new HashSet<string>();
            dfs(s);
            return result;
        }

        private void dfs(string curr)
        {
            //base
            if (curr.Length < max || set.Contains(curr)) return;
            if (IsValid(curr))
            {
                if (curr.Length > max)
                {
                    max = curr.Length;
                    result = new List<string>();
                    result.Add(curr);
                }
                else
                {
                    result.Add(curr);
                }
                set.Add(curr);
            }
            //logic
            set.Add(curr);
            char[] charArray = curr.ToCharArray();
            for (int j = 0; j < curr.Length; j++)
            {
                if (Char.IsLetter(charArray[j])) continue;
                string sub1 = curr.Substring(0, j);
                string sub2 = curr.Substring(j + 1, curr.Length - 1 - (j + 1) + 1);
                string child = sub1 + sub2;

                dfs(child);
            }
        }


    }
}
